Upgrade Antrea base image to ubuntu:22.04

antoninbas opened the pull request:
Ubuntu 20.04 comes with iptables 1.8.4, which seems to have several bugs
when used with nft. In particular, we observe that iptables-restore
sometimes segfaults when restoring the ANTREA-NODE-PORT-LOCAL chain in
the nat table. Ubuntu 22.04 comes with a more recent iptables version, 1.8.7.

As part of this change, we change the tag format for base images
(antrea/openvswitch and antrea/base-ubuntu). We no longer use the OVS
version as the tag, instead we use the Antrea minor version number.

Fixes #4435

Signed-off-by: Antonin Bas <abas@vmware.com>

luolanzone commented on hack/build-antrea-linux-all.sh:
```diff
@@ -112,11 +112,11 @@ GO_VERSION=$(head -n 1 build/images/deps/go-version)
 if $PULL; then
     if [[ ${DOCKER_REGISTRY} == "" ]]; then
-        docker pull $PLATFORM_ARG ubuntu:20.04
+        docker pull $PLATFORM_ARG ubuntu:22.04
```
I searched `20.04`, there are actually a few more files using `ubuntu:20.04`, e.g.:`Dockerfile.octant.ubuntu`, `flow-aggregator/Dockerfile`. Should we upgrade them as well to keep base image consistent? or you feel it's OK to keep using `ubuntu:20.04` if they are not impacted by old iptables?

antoninbas replied:
that's intentional, I'm planning to update the rest of the images in a follow-up PR

antoninbas commented:
/test-all

tnqn reviewed (dismissed):
LGTM

antoninbas commented:
/test-e2e

antoninbas commented:
@tnqn @luolanzone could you take another look?
I realized that if we wanted to keep using the same Ubuntu version for older Antrea versions, we would need to add the Ubuntu version to the tag for `antrea/openvswitch` and `antrea/base-ubuntu`. For example: `antrea/openvswitch:20.04-ovs2.17.3` and `antrea/base-ubuntu:20.04-ovs2.17.3`. Instead of `antrea/openvswitch:2.17.3` and `antrea/base-ubuntu:2.17.3`.

tnqn reviewed (dismissed):
LGTM

tnqn commented:
> I realized that if we wanted to keep using the same Ubuntu version for older Antrea versions, we would need to add the Ubuntu version to the tag for `antrea/openvswitch` and `antrea/base-ubuntu`.

We could use a more specific tag but I assume we need to backport this to older Antrea versions? Otherwise runner's ubuntu version needs to be pinned and user would meet the issue when running Antrea on Jammy if they don't upgrade to 1.10.

tnqn commented:
/test-all

luolanzone commented on build/images/ovs/build.sh:
```diff
@@ -92,6 +92,16 @@ pushd $THIS_DIR > /dev/null
+    "ubi")
+        BUILD_TAG="ovs$OVS_VERSION"
+        ;;
+esac
```
Better to add a line here as well to print the BUILD_TAG:
`echo "BUILD_TAG: $BUILD_TAG"`

antoninbas replied:
done

luolanzone reviewed (approved):
LGTM overall, one nit. And you may need to update the PR's summary as well.

antoninbas commented:
I have squashed commits

antoninbas commented:
/test-all

tnqn commented:
> By "more specific tag", I assume you mean something based on the current Antrea version?

By "more specific tag" I meant the tag including "22.04" which is already added by the PR.

tnqn commented:
"Go / Verify docs and spelling" failed because there is a stale link about update_ovs_image.yml which doesn't exist.

tnqn commented:
@antoninbas while I'm working on #4484 which adds suricata to the base images, I realize that it's not extensible to name the tag in the current way for long term. For example, once #4484 is merged, the previous releases built after it would include the bits unnecessarily (if this PR is backported). However, it would be also be weird if we continue adding suricata version to the tag (which would become base-ubuntu:20.04-ovs2.17.3-suricata6.0).
I'm thinking if we should just name the tag with numbers, and perhaps add a VERSION file to the image directory like https://github.com/kubernetes/kubernetes/blob/master/test/images/agnhost/VERSION and increase it every time the image is updated.

antoninbas commented:
@tnqn I changed the tag format, it is now `22.04_v1.10_f26634ec4c`.

I just remembered as I was working on this that **we always build all base images from scratch when we generate an image for a patch release**. I introduced this mechanism a long time ago to avoid such issues. In other words, we only push base images to the Docker registry when we build the "latest" Antrea image, and this is for caching purposes only.

I still decided to update the tag, as I don't think it makes sense to keep using the OVS version as the tag forever. In theory we could just use `latest` since it is for caching purposes only, but maybe it's not that great.

The reason for including the Antrea version is if we want to identify and delete old images from the registry. 

If this is merged, to add Suricata please:
1) add `build/images/deps/suricata-version` with the correct version number
2) update `build/images/chains/ubi.def` and `build/images/chains/ubuntu.def` to indicate that Suricata is included in the image

antoninbas commented:
/test-all

tnqn commented:
/test-latest-all

tnqn commented:
/test-latest-all

tnqn commented:
> @tnqn I changed the tag format, it is now `22.04_v1.10_f26634ec4c`.
> 
> I just remembered as I was working on this that **we always build all base images from scratch when we generate an image for a patch release**. I introduced this mechanism a long time ago to avoid such issues. In other words, we only push base images to the Docker registry when we build the "latest" Antrea image, and this is for caching purposes only.
> 
> I still decided to update the tag, as I don't think it makes sense to keep using the OVS version as the tag forever. In theory we could just use `latest` since it is for caching purposes only, but maybe it's not that great.
> 
> The reason for including the Antrea version is if we want to identify and delete old images from the registry.
> 
> If this is merged, to add Suricata please:
> 
> 1. add `build/images/deps/suricata-version` with the correct version number
> 2. update `build/images/chains/ubi.def` and `build/images/chains/ubuntu.def` to indicate that Suricata is included in the image

@antoninbas thanks for the explanation. It also reminds me the original purpose of the base images. Then does it make sense to just use the major + minor parts of antrea VERSION as the tag? Then there will be a "latest" image for each release and we don't need to care about the changes of dependencies and the base images generated during the development cycle. I see it's also a little difficult to predict the image tag in various scripts, for instance, the tag used in `build/images/test/Dockerfile`. Using a stable tag should resolve it.

But to avoid confusion caused by the tag, "openvswitch:1.10", perhaps we could have some prefix to indicate it's not OVS version, like "openvswitch:antrea-1.10", "base-ubuntu:antrea-1.10"?

antoninbas commented:
@tnqn the simpler approach works for me. I changed the tag to `antrea-v1.10`. It helps to have the same tag for both build chains (ubuntu and ubi).

antoninbas commented:
/test-latest-all

antoninbas commented:
/test-latest-all

tnqn commented:
/test-conformance

tnqn reviewed (approved):
LGTM, thanks
